enhancements/storage/csi-driver-operator-merge.md

```diff
@@ -0,0 +1,551 @@
+manually (or with some tricks in CI if possible, see below).
+
+Once we're happy with the results, we will follow ART's guidelines and flip building of an existing image from the old
+repo to github.com/openshift/csi-operator. The resulting image name will be the same. This will require a tight
+cooperation with ART. We expect guidelines
+like [Changing the component name of a second level operator](https://docs.ci.openshift.org/docs/how-tos/onboarding-a-new-component/#changing-the-component-name-of-a-second-level-operator) -
+we're not changing name of an operator, but we want to change the source repository of the image.
+
+Example:
+
+* In OCP 4.14, ART builds
+  only [`ose-aws-ebs-csi-driver-operator-container`](https://brewweb.engineering.redhat.com/brew/packageinfo?packageID=74505)
+  from https://github.com/openshift/aws-ebs-csi-driver-operator repository.
+
+* In 4.15, we merge the operator into github.com/openshift/csi-operator into `/legacy` directory, including
+  its own `go.mod` and `vendor/`, and do not change any code of it. This way, we can ensure the code will be 100% the
+  same as the
+  old operator
+  * We can test the new image pre-merge manually using:
+    `oc adm release new --from=<the latest 4.15 nightly> aws-ebs-csi-driver-operator=quay.io/jsafrane/my-ebs-operator:1 --to=quay.io/jsafrane/test-release:1` and install / upgrade from `test-release:1`.
+  * _We could trick CI to build `aws-ebs-csi-driver-operator` image from the new repo and use it in presubmit
+    tests there. But we can't promote it anywhere, as it would overwrite the old image from
+    github.com/openshift/aws-ebs-csi-driver-operator. Some experiments are needed here._
+
+  The openshift/csi-operator repository should look like this at this point:
+  ```text
+  ├── assets          # Common assets for all CSI drivers (empty now)
+  ├── cmd             # All commands for all CSI driver operators (empty now)
+  ├── pkg             # All common code for all CSI driver operators (empty now)
+  ├── test            # All tests for all CSI driver operators (empty now)
+  ├── vendor          # Global vendor dir (empty now)
+  ├── legacy          # "Old" CSI driver operators that just merged here.
+  │   └── aws-ebs-csi-driver-operator
+  │       ├── assets
+  │       │   ├── assets.go
+  │       │   ├── cabundle_cm.yaml
+  │       │   └── ... # all other assets
+  │       ├── cmd
+  │       │   └── aws-ebs-csi-driver-operator
+  │       │       └── main.go
+  │       ├── pkg
+  │       │   ├── dependencymagnet
+  │       │   │   └── dependencymagnet.go
+  │       │   ├── operator
+  │       │   │   ├── starter.go
+  │       │   │   ├── starter_test.go
+  │       │   │   ├── storageclasshook.go
+  │       │   │   └── storageclasshook_test.go
+  │       │   └── version
+  │       │       └── version.go
+  │       ├── test
+  │       │   └── e2e
+  │       │       └── manifest.yaml
+  │       ├── vendor
```

Comment:
We do need to take care cachito is looking at the right things to make the build happen. 

*Option 1*
Production builds will need to get config to only look at `legacy/image-name` directory during the transition period (with `content.source.path:`), which means that the Dockerfile need to be consumed from that directory. 
Prior art is here: https://github.com/openshift-eng/ocp-build-data/blob/openshift-4.15/images/vertical-pod-autoscaler.yml

*Option 2*
Use `WORKDIR` in Dockerfile during the transition period to confine what the build environment looks like during the transition period, and configure `cachito.packages.gomod[].path` to the directory where `go.mod` and `vendor/` can be found

Reply:
Can you clarify where we could update the cachito's configuration?

Reply:
Cachito configuration lives in ocp-build-data and can be worked on with ART.